DEV-364 Create new config object and YAML container

moseshll opened the pull request:
- `Config.pm` module for reading YML config files.
- x.yml/x.local.yml/ENV hierarchy for Apache and k8s support.
- New `lib/CRMS` directory for modules to migrate into.
- Config variable cleanup: limited renaming plus camel -> snake case.
- `bin/legacy` binaries jettisoned.
- Correct `MARC::Record::MiJ` spelling.
- (`cgi/crms`) Prettify landing page debug with CSS and `Data::Dumper` for better config display.
- Use upstream `wait-for` and deal intelligently with spawning `crms` and `ht` (`db-image`) databases.
- Remove dependency on CRMS module from Jira and Data API modules since they can now use config directly.

Caveat:
The last two commits are in support of running reports under development against the production DB, i.e., having both hosts available in the config file and choosing one at runtime. Long-standing CRMS patterns unfortunately keep this flexibility as part of app logic rather than just config.

Things to look at:
- `lib/CRMS/Config.pm`
- Changes to `cgi/CRMS.pm` in vicinity of line 350 with database connections.

Things to ignore:
- Deleted files
- Renamed/snake-cased config keys (mainly in .tt files and 'bin/` cron scripts).

moseshll committed:
DEV-364 Create new config object and YAML container
- Config.pm module for reading YML config files.
- x.yml/x.local.yml/ENV hierarchy for Apache and k8s support.
- New lib/CRMS directory for modules to migrate into.
- Config variable cleanup: limited renaming plus camel -> snake case.
- bin/legacy binaries jettisoned.
- Correct MARC::Record::MiJ spelling.
- Prettify landing page debug with CSS and Data::Dumper for better config display.
- Use upstream wait-for and deal intelligently with spawning crms and ht (db-image) databases.

aelkiss commented on scripts/test_and_cover.sh:
```diff
@@ -1,4 +1,4 @@
 #!/bin/bash
 
-bin/wait-for --timeout=300 mariadb:3306
+bin/wait-for --timeout=300 mariadb:3306 mariadb_ht:3306
```
```bash
bin/wait-for --timeout=300 mariadb:3306
# or set lower timeout for the second
bin/wait-for --timeout=300 mariadb_ht:3306
```

moseshll replied:
Using upstream `wait-for` with a few additional tweaks seems to be working.

moseshll commented:
~~Need to check whether the `bin/reminder.pl` cron and the `disableReminder` config variable are still needed.~~

moseshll commented:
~~Also TODO: clean up comments in `config/config.yml` and add any additional documentation in `lib/CRMS/Config.pm`.~~

moseshll commented:
Things to look at:
- `lib/CRMS/Config.pm`
- Changes to `cgi/CRMS.pm` in particular vicinity of line 350 with database connections.
Things to ignore:
- Deleted files
- Renamed/snake-cased config keys (mainly in `.tt` files and 'bin/` cron scripts).

moseshll commented:
Reverted to draft, since `Jira.pm` routines no longer need a `$crms` parameter passed to every routine. Those can be eliminated.

billdueber commented on bib_rights/import_bib_rights_data.pl:
```diff
@@ -16,7 +16,7 @@ BEGIN
 use IO::Zlib;
 use MARC::Record;
-use MARC::Record::MIJ;
+use MARC::Record::MiJ;
```
Hey! I wrote that!

moseshll replied:
A calculated risk. Pretty sketchy stuff be be importing into our nice apps ;-)

billdueber commented on cgi/CRMS.pm:
```diff
@@ -4,20 +4,30 @@ package CRMS;
 ## Object of shared code for the CRMS DB CGI and BIN scripts.
 ## ----------------------------------------------------------
 
+BEGIN {
```
Do you want to check to make sure ENV{SDRROOT} is set up here? I imagine some of these `use`s will fail otherwise.

moseshll replied:
I don't think I've ever had a sanity check for SDRROOT or DLXSROOT for that matter. Makes sense.

billdueber commented on cgi/CRMS.pm:
```diff
@@ -404,39 +359,29 @@ sub DbInfo
-    $sdr_dbh->{mysql_enable_utf8} = 1;
-    $sdr_dbh->do('SET NAMES "utf8";');
+  my $self    = shift;
+  my $db_name = shift || 'ht';
```
Worth putting the default `ht` into the config object?

moseshll replied:
Yes, especially in light of the fact that `$db_name` as a parameter will probably go away.

billdueber commented on cgi/CRMS.pm:
```diff
@@ -454,15 +399,16 @@ sub GetSdrDb
   return $sdr_dbh;
```
Is there a difference between `$sdr_dbh` and `$ht_dbh` in the code? It kinda looked like you were trying to rename the latter to the former.

moseshll replied:
All things SDR will be renamed HT, but I'm gonna confine most of that to a separate cleanup ticket that's still just over the horizon: a stand-alone DB module now that we have pretty much got stand-alone config in the can.

billdueber commented on cgi/CRMS.pm:
```diff
@@ -454,15 +399,16 @@ sub GetSdrDb
+  my $db_host = $config->config->{'db_host'};
+  $db_host = $config->config->{'db_host_development'} if $instance eq '';
+  my $db_name = 'crms';
+  $db_name = 'crms_training' if $instance eq 'crms-training';
```
Should the dev info also go in the config object as opposed to being hardcoded here?

moseshll replied:
For consistency with previous comment, yes to storing defaults in the config.

billdueber reviewed (approved):
I won't even pretend to have followed all the logic, but the changes all seem internally consistent. Just a couple notes about maybe putting more stuff into the config object.
